pkg/ddl/executor.go

```diff
@@ -886,34 +860,91 @@ func GetDefaultCollation(cs string, defaultUTF8MB4Collation string) (string, err
 }
 
 // ResolveCharsetCollation will resolve the charset and collate by the order of parameters:
-// * If any given ast.CharsetOpt is not empty, the resolved charset and collate will be returned.
-// * If all ast.CharsetOpts are empty, the default charset and collate will be returned.
+//
+// charsetOpts[0] == explicitly set charset/collation for the object (object can be a table, database, etc)
+// charsetOpts[1] == Template object charset/collation. For example a CREATE TABLE uses the charset/collation
+// of the schema as template.
 func ResolveCharsetCollation(charsetOpts []ast.CharsetOpt, utf8MB4DefaultColl string) (chs string, coll string, err error) {
-	for _, v := range charsetOpts {
-		if v.Col != "" {
-			collation, err := collate.GetCollationByName(v.Col)
+	if len(charsetOpts) == 0 {
+		chs, _ = charset.GetDefaultCharsetAndCollate()
+		coll, err = GetDefaultCollation(chs, utf8MB4DefaultColl)
+		if err != nil {
+			return "", "", errors.Trace(err)
+		}
+		return
+	}
+
+	// Handle explicit collation and charset for the object
+	if len(charsetOpts) > 0 {
+		if charsetOpts[0].Chs != "" {
+			info, err := charset.GetCharsetInfo(charsetOpts[0].Chs)
+			if err != nil {
+				return "", "", errors.Trace(err)
+			}
+			chs = info.Name
```

Comment:
from the rule of MySQL, collation must be the default of the specified charset in this case https://dev.mysql.com/doc/refman/8.4/en/charset-table.html

this code cannot handle this case(not sure whether the previous one can handle it)

```
CREATE SCHEMA s1 DEFAULT CHARACTER SET utf8 COLLATE utf8_unicode_ci;
USE s1;
CREATE TABLE t11 (id INT PRIMARY KEY);
CREATE TABLE t12 (id INT PRIMARY KEY) CHARACTER SET utf8;
SELECT TABLE_NAME, TABLE_COLLATION FROM information_schema.tables WHERE TABLE_SCHEMA='s1';
```

Reply:
Current master:
```
mysql-8.0.11-TiDB-v9.0.0-alpha-95-gc44672e0c5> SELECT TABLE_NAME, TABLE_COLLATION FROM information_schema.tables WHERE TABLE_SCHEMA='s1';
+------------+-----------------+
| TABLE_NAME | TABLE_COLLATION |
+------------+-----------------+
| t11        | utf8_unicode_ci |
| t12        | utf8_bin        |
+------------+-----------------+
2 rows in set (0.00 sec)
```

Note that `utf8_bin` is the default for `utf8` in TiDB.